### PR TITLE
Docker-oriented Makefile

### PR DESCRIPTION
**Pull request by ailispaw**
I gave up my previous idea with two Dockerfiles.

I made something Docker-oriented Makefile.
It's ugly, but it works.
I hope it's helpful for #3 .

### CODE REVIEW

**Comment by ailispaw** [comment]
I will make sure timezone of `stat` and `date`.

**Comment by ailispaw** [comment]
It's not good enough yet in some cases. I will figure it out.
A bash script seems much easier than Makefile in this case though.

File: Makefile [hunk 1]
```diff
@@ -1,23 +1,46 @@
...
-build: | ccache dl
-	docker build --no-cache -t dhyve-os-build .
-	-docker rm dhyve-os-built
-	docker run -v ${PWD}/ccache:/build/buildroot/ccache -v ${PWD}/dl:/build/buildroot/dl --name=dhyve-os-built dhyve-os-build
+BUILT := `docker ps -aq -f name=$(BUILD_CONTAINER) -f exited=0`
+
+all: $(TARGETS)
+
+$(TARGETS): build | output
+	@if [ "$(BUILT)" != "" ]; then \
```
Comment (nlf) [inline Makefile]: if this target depends on build, is this if statement necessary?

File: Makefile [hunk 2]
```diff
@@ -1,23 +1,46 @@
...
+		echo docker cp $(BUILD_CONTAINER):/build/buildroot/output/images/$(@F) output/; \
```
Comment (nlf) [inline Makefile]: why the echo?

File: Makefile [hunk 3]
```diff
@@ -1,23 +1,46 @@
-all: output/rootfs.cpio.xz output/bzImage
+TARGETS := output/rootfs.cpio.xz output/bzImage
+SOURCES := Dockerfile config/buildroot config/kernel config/user bin/build_dhyve
```
Comment (nlf) [inline Makefile]: hmm.. will this trigger a rebuild if something changes in the rootfs directory too?

File: Makefile (hunk 2 above)
Comment (ailispaw) [inline Makefile]: Just I would like to see progresses.
They are unnecessary. You can remove them.

File: Makefile (hunk 1 above)
Comment (ailispaw) [inline Makefile]: Hmm. It seems not.
I thought it's necessary in some case at the first time.

File: Makefile [hunk 4]
```diff
@@ -1,23 +1,46 @@
...
 
-build: | ccache dl
-	docker build --no-cache -t dhyve-os-build .
-	-docker rm dhyve-os-built
-	docker run -v ${PWD}/ccache:/build/buildroot/ccache -v ${PWD}/dl:/build/buildroot/dl --name=dhyve-os-built dhyve-os-build
+BUILT := `docker ps -aq -f name=$(BUILD_CONTAINER) -f exited=0`
+
+all: $(TARGETS)
+
+$(TARGETS): build | output
```
Comment (ailispaw) [inline Makefile:21]: I'm not sure if build should be an order-only-prerequisite.

Comment (ailispaw) [inline Makefile:21]: Confirmed. It should be a normal-prerequisite.

File: Makefile (hunk 3 above)
Comment (ailispaw) [inline Makefile]: Oops. I missed them.

**Comment by ailispaw** [comment]
Fixed.

**Comment by ailispaw** [comment]
May I squash?

**Comment by nlf** [comment]
Absolutely

**Comment by ailispaw** [comment]
Thanks.

But I found another bug that it won't catch `docker build` error inside if statement.

So can I `docker rmi $(BUILD_IMAGE)` first?
And also we have to check `$(BUILT)` in the `$(TARGETS)` recipe in this case.

**Comment by ailispaw** [comment]
It needs `set -e` inside if statements.

**Comment by nlf** [comment]
Ah cool, you figured it out :)

**Comment by nlf** [comment]
merging this now, it looks good. thanks!

**Comment by ailispaw** [comment]
Thanks for this opportunity.
It was a fun challenging to play with Makefile and Docker.

**Pull request merged**